File: Polars.CSharp/Expr/ArrayOps.cs
using Polars.NET.Core;
using Pl = Polars.CSharp.Polars;

namespace Polars.CSharp;

// ==========================================
// ArrayOps Helper Class
// ==========================================
/// <summary>
/// Offers methods for accessing fields within array columns.
/// </summary>
public readonly struct ArrayOps
{
    private readonly Expr _expr;
    internal ArrayOps(Expr expr) { _expr = expr; }

    private Expr Wrap(Func<ExprHandle, ExprHandle> op)
        => new(op(_expr.CloneHandle()));
    /// <inheritdoc cref="SeriesArrayOps.Len"/>
    public Expr Len() => Wrap(PolarsWrapper.ArrayLen);
    /// <inheritdoc cref="SeriesArrayOps.Max"/>
    public Expr Max() => Wrap(PolarsWrapper.ArrayMax);
    /// <inheritdoc cref="SeriesArrayOps.Min"/>
    public Expr Min() => Wrap(PolarsWrapper.ArrayMin);
    /// <inheritdoc cref="SeriesArrayOps.Sum"/>
    public Expr Sum() => Wrap(PolarsWrapper.ArraySum);
    /// <inheritdoc cref="SeriesArrayOps.Mean"/>
    public Expr Mean() => Wrap(PolarsWrapper.ArrayMean); 
    /// <inheritdoc cref="SeriesArrayOps.Median"/>
    public Expr Median() => Wrap(PolarsWrapper.ArrayMedian); 
    /// <inheritdoc cref="SeriesArrayOps.NUnique"/>
    public Expr NUnique() => Wrap(PolarsWrapper.ArrayNUnique); 
    /// <inheritdoc cref="SeriesArrayOps.Std"/>
    public Expr Std(byte ddof = 1) => new(PolarsWrapper.ArrayStd(PolarsWrapper.CloneExpr(_expr.Handle), ddof)); 
    /// <inheritdoc cref="SeriesArrayOps.Var"/>
    public Expr Var(byte ddof = 1) => new(PolarsWrapper.ArrayVar(PolarsWrapper.CloneExpr(_expr.Handle), ddof)); 
    /// <inheritdoc cref="SeriesArrayOps.CountMatches"/>
    public Expr CountMatches(Expr element) => new(PolarsWrapper.ArrayCountMatches(_expr.CloneHandle(),element.CloneHandle()));
    /// <inheritdoc cref="SeriesArrayOps.Agg"/>
    public Expr Agg(Expr expr) => new(PolarsWrapper.ArrayAgg(_expr.CloneHandle(),expr.CloneHandle()));
    /// <inheritdoc cref="SeriesArrayOps.Eval"/>
    public Expr Eval(Expr expr,bool asList=false) => new(PolarsWrapper.ArrayEval(_expr.CloneHandle(),expr.CloneHandle(),asList));
    /// <inheritdoc cref="SeriesArrayOps.First"/>
    public Expr First() => Get(0,true);
    /// <inheritdoc cref="SeriesArrayOps.Last"/>
    public Expr Last() => Get(-1,true);
    // --- Boolean ---
    /// <inheritdoc cref="SeriesArrayOps.Any"/>
    public Expr Any() => Wrap(PolarsWrapper.ArrayAny); 
    /// <inheritdoc cref="SeriesArrayOps.All"/>
    public Expr All() => Wrap(PolarsWrapper.ArrayAll); 
    /// <inheritdoc cref="SeriesArrayOps.Sort"/>
    public Expr Sort(bool descending = false, bool nullsLast = false, bool maintainOrder = false)
        => new(PolarsWrapper.ArraySort(_expr.CloneHandle(), descending, nullsLast, maintainOrder));
    /// <inheritdoc cref="SeriesArrayOps.Reverse"/>
    public Expr Reverse() => Wrap(PolarsWrapper.ArrayReverse); 
    /// <inheritdoc cref="SeriesArrayOps.ArgMin"/>
    public Expr ArgMin() => Wrap(PolarsWrapper.ArrayArgMin); 
    /// <inheritdoc cref="SeriesArrayOps.ArgMax"/>
    public Expr ArgMax() => Wrap(PolarsWrapper.ArrayArgMax); 
    internal Expr Get(Expr index, bool nullOnOob = true)
        => new(PolarsWrapper.ArrayGet(_expr.CloneHandle(), index.CloneHandle(), nullOnOob));
    /// <inheritdoc cref="SeriesArrayOps.Get"/>
    public Expr Get(long index,bool nullOnOob = true) => Get(Pl.Lit(index),nullOnOob);
    /// <inheritdoc cref="SeriesArrayOps.Join"/>
    public Expr Join(string separator, bool ignoreNulls = true)
        => new(PolarsWrapper.ArrayJoin(_expr.CloneHandle(), separator, ignoreNulls));
    /// <inheritdoc cref="SeriesArrayOps.Explode"/>
    public Expr Explode(bool emptyAsNull=true,bool keepNulls = true)
        => new(PolarsWrapper.ArrayExplode(_expr.CloneHandle(),emptyAsNull,keepNulls)); 
    
    /// <summary>
    /// Convert array to struct.
    /// </summary>
    /// <param name="fields">Desired field names</param>
    public Expr ToStruct(params string[] fields)
        => new(PolarsWrapper.ArrayToStruct(_expr.CloneHandle(),fields)); 
    /// <summary>
    /// Convert array to a struct using a function to generate field names dynamically.
    /// </summary>
    /// <param name="nameGenerator">A function that takes an index and returns a string.</param>
    /// <param name="fieldCount">The expected number of fields to generate.</param>
    public Expr ToStruct(Func<int, string> nameGenerator, int fieldCount)
    {
        if (fieldCount <= 0) 
            return ToStruct();

        string[] fields = new string[fieldCount];
        for (int i = 0; i < fieldCount; i++)
        {
            fields[i] = nameGenerator(i);
        }
        return ToStruct(fields);
    }
    /// <inheritdoc cref="SeriesArrayOps.ToList"/>
    public Expr ToList() => Wrap(PolarsWrapper.ArrayToList);
    /// <inheritdoc cref="SeriesArrayOps.Contains"/>
    public Expr Contains(Expr item, bool nullsEqual = false)
        => new(PolarsWrapper.ArrayContains(_expr.CloneHandle(),item.CloneHandle(), nullsEqual));
    /// <inheritdoc cref="SeriesArrayOps.Unique"/>
    public Expr Unique(bool stable = false)
        => new(PolarsWrapper.ArrayUnique(_expr.CloneHandle(), stable));
    /// <summary>
    /// Concat this Array expression with other Array expressions.
    /// </summary>
    /// <param name="others">Other list expressions to append.</param>
    public Expr Concat(params Expr[] others)
    {
        var allExprs = new ExprHandle[others.Length + 1];

        allExprs[0] = _expr.CloneHandle();

        for (int i = 0; i < others.Length; i++)
        {
            allExprs[i + 1] = others[i].CloneHandle();
        }

        return new Expr(PolarsWrapper.ConcatArray(allExprs));
    }
}

File: Polars.FSharp/Expr/ArrayOps.fs
namespace Polars.FSharp

open System
open Polars.NET.Core

type [<Struct>] ArrayOps(handle: ExprHandle) = 
    member _.Len() = new Expr(PolarsWrapper.ArrayLen handle)
    /// <summary> Compute the sum of the values in the array. </summary>
    member _.Sum() = new Expr(PolarsWrapper.ArraySum handle)

    /// <summary> Compute the minimum value in the array. </summary>
    member _.Min() = new Expr(PolarsWrapper.ArrayMin handle)

    /// <summary> Compute the maximum value in the array. </summary>
    member _.Max() = new Expr(PolarsWrapper.ArrayMax handle)

    /// <summary> Compute the mean value in the array. </summary>
    member _.Mean() = new Expr(PolarsWrapper.ArrayMean handle)

    /// <summary> Compute the median value in the array. </summary>
    member _.Median() = new Expr(PolarsWrapper.ArrayMedian handle)
    /// <summary>
    /// Count the number of unique values in every sub-arrays.
    /// </summary>
    member _.NUnique() = new Expr(PolarsWrapper.ArrayNUnique handle)
    /// <summary> Compute the standard deviation of the values in the array. </summary>
    member _.Std(?ddof: int) = 
        let d = defaultArg ddof 1 |> byte
        new Expr(PolarsWrapper.ArrayStd(handle, d))

    /// <summary> Compute the variance of the values in the array. </summary>
    member _.Var(?ddof: int) = 
        let d = defaultArg ddof 1 |> byte
        new Expr(PolarsWrapper.ArrayVar(handle, d))
    /// <summary>
    /// Count how often the value produced by element occurs.
    /// </summary>
    /// <param name="element">An expression that produces a single value</param>
    member _.CountMatches(element:Expr) = new Expr(PolarsWrapper.ArrayCountMatches(handle,element.CloneHandle()))
    member _.Agg(element:Expr) = new Expr(PolarsWrapper.ArrayAgg(handle,element.CloneHandle()))
    /// <summary>
    /// Run any polars expression against the arrays’ elements.
    /// </summary>
    /// <param name="expr">Expression to run. Note that you can select an element with pl.element()</param>
    /// <param name="asList">Collect the resulting data as a list. This allows for expressions which output a variable amount of data.</param>
    member _.Eval(expr:Expr,?asList:bool) =
        let asL = defaultArg asList false
        new Expr(PolarsWrapper.ArrayEval(handle,expr.CloneHandle(),asL))
    /// <summary> Check if any value in the array is true. </summary>
    member _.Any() = new Expr(PolarsWrapper.ArrayAny handle)

    /// <summary> Check if all values in the array are true. </summary>
    member _.All() = new Expr(PolarsWrapper.ArrayAll handle)

    /// <summary> Check if the array contains a specific item. </summary>
    member _.Contains(item: Expr, ?nullsEqual: bool) =
        let eq = defaultArg nullsEqual false
        // Wrapper consumes item ownership
        new Expr(PolarsWrapper.ArrayContains(handle, item.CloneHandle(), eq))

    /// <summary> Check if the array contains a literal value. </summary>
    member this.Contains(item: string, ?nullsEqual) =
        let eq = defaultArg nullsEqual false 
        let itemHandle = PolarsWrapper.Lit item
        new Expr(PolarsWrapper.ArrayContains(handle, itemHandle, eq))
    member this.Contains(item: int, ?nullsEqual) =
        let eq = defaultArg nullsEqual false 
        let itemHandle = PolarsWrapper.Lit item
        new Expr(PolarsWrapper.ArrayContains(handle, itemHandle, eq))
    // --- Operations ---

    /// <summary> Get unique values in the array. </summary>
    member _.Unique(?stable: bool) = 
        let s = defaultArg stable false
        new Expr(PolarsWrapper.ArrayUnique(handle, s))

    /// <summary> Join array elements into a string. </summary>
    member _.Join(separator: string, ?ignoreNulls: bool) =
        let ign = defaultArg ignoreNulls true
        new Expr(PolarsWrapper.ArrayJoin(handle, separator, ign))

    /// <summary> Sort the array. </summary>
    member _.Sort(?descending: bool, ?nullsLast: bool, ?maintainOrder: bool) =
        let desc = defaultArg descending false
        let nLast = defaultArg nullsLast false
        let stable = defaultArg maintainOrder false 
        new Expr(PolarsWrapper.ArraySort(handle, desc, nLast, stable))

    /// <summary> Reverse the array. </summary>
    member _.Reverse() = new Expr(PolarsWrapper.ArrayReverse handle)

    /// <summary> Get the index of the minimum value. </summary>
    member _.ArgMin() = new Expr(PolarsWrapper.ArrayArgMin handle)

    /// <summary> Get the index of the maximum value. </summary>
    member _.ArgMax() = new Expr(PolarsWrapper.ArrayArgMax handle)

    /// <summary> Explode the array to rows. </summary>
    member _.Explode(?emptyAsNull:bool,?keepNulls:bool) = 
        let emp = defaultArg emptyAsNull true
        let kn = defaultArg keepNulls true
        new Expr(PolarsWrapper.ArrayExplode(handle,emp,kn))

    /// <summary> Get value at integer index. </summary>
    member this.Get(index: int, ?nullOnOob) =
        let oob = defaultArg nullOnOob false
        let indexHandle = PolarsWrapper.Lit index
        new Expr(PolarsWrapper.ArrayGet(handle, indexHandle, oob))
    /// <summary>
    /// Get the first value of the sub-arrays.
    /// </summary>
    member this.First() = this.Get(0,true)
    /// <summary>
    /// Get the last value of the sub-arrays.
    /// </summary>
    member this.Last() = this.Get(-1,true)

    /// <summary>
    /// Combine the current expression with other expressions into an Array.
    /// Result: [parent_val, other_val_1, other_val_2, ...]
    /// Equivalent to: pl.concatArray([parent, others...])
    /// </summary>
    member _.Concat(others: seq<#IColumnExpr>) =
        let currentHandle = handle
        let handles = 
            seq {
                yield currentHandle
                
                yield! others 
                       |> Seq.collect (fun x -> x.ToExprs()) 
                       |> Seq.map (fun e -> e.CloneHandle())
            }
            |> Seq.toArray

        new Expr(PolarsWrapper.ConcatArray handles)

    // --- Conversion ---

    /// <summary> Convert Array to List (variable length). </summary>
    member _.ToList() = new Expr(PolarsWrapper.ArrayToList handle)

    /// <summary> Convert Array to Struct. </summary>
    member _.ToStruct([<ParamArray>]names:string array) = new Expr(PolarsWrapper.ArrayToStruct(handle,names))